core/trino-main/src/main/java/io/trino/execution/scheduler/StageEventDrivenTaskSourceFactory.java

```diff
@@ -0,0 +1,1310 @@
+
+        private Map<Integer, TaskPartition> outputPartitionToTaskPartition;
+        private final Set<Integer> createdTaskPartitions = new HashSet<>();
+        private final Set<PlanNodeId> completedSources = new HashSet<>();
+        private final ListMultimap<PlanNodeId, Split> replicatedSplits = ArrayListMultimap.create();
+
+        private int nextTaskPartitionId;
+
+        HashDistributionSplitAssigner(
+                Optional<CatalogHandle> catalogRequirement,
+                Set<PlanNodeId> partitionedSources,
+                Set<PlanNodeId> replicatedSources,
+                long targetPartitionSizeInBytes,
+                Map<PlanNodeId, OutputDataSizeEstimate> outputDataSizeEstimates,
+                FaultTolerantPartitioningScheme sourcePartitioningScheme,
+                boolean preserveOutputPartitioning)
+        {
+            this.catalogRequirement = requireNonNull(catalogRequirement, "catalogRequirement is null");
+            this.partitionedSources = ImmutableSet.copyOf(requireNonNull(partitionedSources, "partitionedSources is null"));
+            this.replicatedSources = ImmutableSet.copyOf(requireNonNull(replicatedSources, "replicatedSources is null"));
+            allSources = ImmutableSet.<PlanNodeId>builder()
+                    .addAll(partitionedSources)
+                    .addAll(replicatedSources)
+                    .build();
+            this.targetPartitionSizeInBytes = targetPartitionSizeInBytes;
+            this.outputDataSizeEstimates = ImmutableMap.copyOf(requireNonNull(outputDataSizeEstimates, "outputDataSizeEstimates is null"));
+            this.sourcePartitioningScheme = requireNonNull(sourcePartitioningScheme, "sourcePartitioningScheme is null");
+            this.preserveOutputPartitioning = preserveOutputPartitioning;
+        }
+
+        @Override
+        public AssignmentResult assign(PlanNodeId planNodeId, ListMultimap<Integer, Split> splits, boolean noMoreSplits)
+        {
+            AssignmentResult.Builder assignment = AssignmentResult.builder();
+            if (outputPartitionToTaskPartition == null) {
+                outputPartitionToTaskPartition = createOutputPartitionToTaskPartition();
+            }
+
+            if (replicatedSources.contains(planNodeId)) {
+                replicatedSplits.putAll(planNodeId, splits.values());
+                for (Integer partitionId : createdTaskPartitions) {
+                    assignment.updatePartition(new PartitionUpdate(partitionId, planNodeId, ImmutableList.copyOf(splits.values()), noMoreSplits));
+                }
+            }
+            else {
+                for (Integer outputPartitionId : splits.keySet()) {
+                    TaskPartition taskPartition = outputPartitionToTaskPartition.get(outputPartitionId);
+                    verify(taskPartition != null, "taskPartition not found for outputPartitionId: %s", outputPartitionId);
+                    if (!taskPartition.isIdAssigned()) {
```

Comment:
I wanted to avoid confusing gaps in task ids, as there's always a chance that splits might be missing for a certain task (for example for bucketed tables). Thus I'm trying to assign partition ids lazily. Added a comment.